### DISCUSSION TITLE
Implementation of a way to post-process products

### DISCUSSION THREAD

**Discussion by jfaldanam** [KhaosResearch/greensenti#16] [Ideas] [answered]
In out current use cases, we want to upload to minio and mongo the product files and the metadata. With our current download implementation, we would need to wait until all products have been download to upload them to minio for example.

This can be an issue if we download products in great quantites, as the download host must store all those products in its local disk until the download is completed.

A feature that could help avoid this problem is adding a new optional parameter of type `Callable[pathlib.Path, pandas.DataFrame]` that will post-process each product. This provides a way to compute indexes or crop directly after the download for each product or to solve our usecase upload the product to minio and delete it from local disk.

### Proposed API
```python
def download(
    geojson: Path = None,
    text_match: str = "*",
    from_date: Union[str, datetime] = None,
    to_date: Union[str, datetime] = None,
    *,
    unzip: bool = False,
    max_clouds: int = 100,
    output: Path = Path("."),
    post_process: Callable[Path, pd.DataFrame]
    dhus_username: str = os.environ.get("DHUS_USERNAME", None),
    dhus_password: str = os.environ.get("DHUS_PASSWORD", None),
    dhus_host: str = os.environ.get("DHUS_HOST", None),
    gcloud: Path = os.environ.get("GOOGLE_APPLICATION_CREDENTIALS", None),
) -> pd.DataFrame:
   ...
```

**Comment by benhid** [comment, marked as answer]
I get the use case but am unsure about the proposed implementation. I think the "Pythonic" way should be to `yield` the product instead of returning a data frame. 

Both solutions might require some changes, though. Does the [`copernicous_download()`](https://github.com/KhaosResearch/greensenti/blob/main/src/greensenti/dhus.py#LL230C5-L230C25) function process each product individually?

**Reply by jfaldanam** [reply]
I agree on using the pythonic way, good call, the return type of the function will be a `Generator[pandas.Series]`.


[`copernicous_download()`](https://github.com/KhaosResearch/greensenti/blob/main/src/greensenti/dhus.py#LL230C5-L230C25) uses the [`download_all()`](https://github.com/sentinelsat/sentinelsat/blob/7fd4122cdf85df651ba41d843ae17a6697106483/sentinelsat/download.py#L234) function from [`sentinelsat`](https://github.com/sentinelsat/sentinelsat). After inspecting the code, that function does some fancy multithreading stuff, calling inside the regular [`download()`](https://github.com/sentinelsat/sentinelsat/blob/7fd4122cdf85df651ba41d843ae17a6697106483/sentinelsat/download.py#L105). I believe the ability to process each product independently, for example to save disk space uploading to external storage, outweights the benefits of the multithreading that `sentinelsat` provides.

**Reply by jfaldanam** [reply]
As a additional note, while working on this I just noticed that we were only using one concurrent download anyway.

https://github.com/KhaosResearch/greensenti/blob/8ba31c803d38d4d52ed6a838573cc192afb40471/src/greensenti/dhus.py#L239-L241
